observability/micrometer/src/main/java/org/wildfly/extension/micrometer/metrics/WildFlyMetric.java

```diff
@@ -0,0 +1,106 @@
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.wildfly.extension.micrometer.metrics;
+
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.FAILURE_DESCRIPTION;
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.NAME;
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.OP;
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.OP_ADDR;
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.READ_ATTRIBUTE_OPERATION;
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.RESULT;
+import static org.wildfly.extension.micrometer.MicrometerExtensionLogger.MICROMETER_LOGGER;
+
+import java.util.OptionalDouble;
+
+import org.jboss.as.controller.LocalModelControllerClient;
+import org.jboss.as.controller.PathAddress;
+import org.jboss.as.controller.descriptions.ModelDescriptionConstants;
+import org.jboss.dmr.ModelNode;
+
+public class WildFlyMetric implements Metric {
+
+    private static final ModelNode UNDEFINED = new ModelNode();
+
+    private LocalModelControllerClient modelControllerClient;
+    private PathAddress address;
+    private String attributeName;
+
+    static {
+        UNDEFINED.protect();
+    }
+
+    public WildFlyMetric() {
+
+    }
+
+    public WildFlyMetric(LocalModelControllerClient modelControllerClient, PathAddress address, String attributeName) {
+        this.modelControllerClient = modelControllerClient;
+        this.address = address;
+        this.attributeName = attributeName;
+    }
+
+    public PathAddress getAddress() {
+        return address;
+    }
+
+    @Override
+    public OptionalDouble getValue() {
+        ModelNode result = readAttributeValue(address, attributeName);
+        if (result.isDefined()) {
+            try {
+                return OptionalDouble.of(result.asDouble());
+            } catch (Exception e) {
+                MICROMETER_LOGGER.unableToConvertAttribute(attributeName, address, e);
+            }
+        }
+        return OptionalDouble.empty();
+    }
+
+    private ModelNode readAttributeValue(PathAddress address, String attributeName) {
+        final ModelNode readAttributeOp = new ModelNode();
+        readAttributeOp.get(OP).set(READ_ATTRIBUTE_OPERATION);
+        readAttributeOp.get(OP_ADDR).set(address.toModelNode());
+        readAttributeOp.get(ModelDescriptionConstants.INCLUDE_UNDEFINED_METRIC_VALUES).set(false);
+        readAttributeOp.get(NAME).set(attributeName);
+        ModelNode response = modelControllerClient.execute(readAttributeOp);
+        String error = getFailureDescription(response);
+        if (error != null) {
+            // [WFLY-11933] if the value can not be read if the management resource is not accessible due to RBAC,
+            // it is logged it at a lower level.
+            if (error.contains("WFLYCTL0216")) {
```

Comment:
I'm not sure I have a better idea, but this looks fragile to me.

Reply:
It may be, but it's cloned from here: metrics/src/main/java/org/wildfly/extension/metrics/WildFlyMetric.java

Given that we couldn't get fixes upstream, we had to do this. It *may* not be necessary with the Micrometer impl. I'll follow up on another JIRA.